networking-calico/networking_calico/plugins/ml2/drivers/calico/endpoints.py

```diff
@@ -295,25 +299,31 @@ def add_port_project_data(self, port, context):
             port[PORT_KEY_PROJ_DATA] = proj_data
             return
 
+        # Not cached, so look up the port's project in the Keystone DB.
+        self.cache_port_project_data()
```

Comment:
I think this means we get the penalty of calling `self.keystone.projects.list()` every time, doesn't it?  Perhaps I am not understanding exactly where the penalty concern is.

Reply:
See the condition immediately above this line -- we (still) only do this when we miss in the cache.

This PR strictly avoids potentially having to fill the cache **during**, e.g., the first Neutron port update upon a wSGI webserver start by forcing the cache fill to be performed as part of mechanism driver initialisation. The cache fill logic was ripped out of this section and put into a function so that it could be called both here, where it previously resided, and in the constructor, for that purpose.

There still is the problem where a new project is created at runtime and we perform a port request associated with that project, then we have a fill operation.... that is probably best fixed separately.

Reply:
Sorry, of course you are right, and thanks for explaining further.